[PFW] Integrate PFW algorithm for the LASSO problem to Pycsou

AdriaJ opened the pull request:
The files are provided in a new branch "v2-fw" that comes from the branch "v2-dev" and thus is up to date with that latter.

Points to be discussed:
* Acceleration strategies for the reweighting step of PFW when the restricted support lasso sub-problem can be solved more efficiently
* Multidimensional inputs/outputs. As the current implementation of PFW does not take any array as input (no "x0" provided), it does not make sense to provide multidimensional handling, all the returned solutions would be identical. We could change this behavior by allowing an optional starting point x0 as input.
* The algorithm has not yet been tested with dask array nor user-defined precision but it should be compliant and able to handle them.

AdriaJ commented:
Reviewers:
@SepandKashani @matthieumeo @joanrue

joanrue commented on src/pycsou/_dev/__init__.py:
```diff
@@ -3,23 +3,37 @@
+        self._lipschitz = np.inf
         self._diff_lipschitz = 2
 
-    def apply(self, arr: pyct.NDArray) -> pyct.Real:
```
Some of the methods have type hints while others don't. I'd suggest adding them again for consistency.

joanrue commented on src/pycsou/_dev/fw_utils.py:
```diff
@@ -0,0 +1,122 @@
+import pycsou.util.ptype as pyct
+
+
+class SquaredL2Norm(pyco.ProxDiffFunc):
```
Same comment as for _dev.SquaredL2Norm

joanrue commented on src/pycsou/opt/solver/fw_lasso.py:
```diff
@@ -0,0 +1,463 @@
+    def m_init(self, **kwargs):
+        super(VanillaFWforLasso, self).m_init(**kwargs)
+        self._mstate["lift_variable"] = 0.0  # todo [0]
+        # QfR: Do I need to enforce the precision on plain floats ?
```
From what I recall, we enforce precision of plain numbers as well --> 
`pycrt.coerce(0.0)`

joanrue commented on src/pycsou/opt/solver/fw_lasso.py:
```diff
@@ -0,0 +1,463 @@
+        new_ind = xp.argmax(xp.abs(mgrad), axis=-1)
+        dcv = mgrad[new_ind] / self.lambda_  # todo [0]
+        mst["dcv"] = dcv
+        # To be careful about mst["dcv"] as it can be either float if new_ind is size 1 or NDArray if new_ind is size>1
```
I am not sure if this is an issue, but in case you wanted to always have NDArray you could use: 
```
new_ind = xp.empty(mgrad.shape[:-1], dtype=int32)
xp.argmax(xp.abs(mgrad), axis=-1, out=new_ind)
```

joanrue commented on src/pycsou/_dev/__init__.py:
```diff
@@ -3,23 +3,37 @@
+        return y
+
+    #    @pycrt.enforce_precision(i=["data"])
+    def asloss(self, data: typ.Optional[pyct.NDArray] = None) -> pyco.ProxFunc:
```
This was my mistake, the base class uses `as_loss`, with **underscore**. We should use the same method name.

joanrue commented on src/pycsou/_dev/example_fw.py:
```diff
@@ -0,0 +1,164 @@
+    print("\tSolved in {:.3f} seconds".format(time_p))
+
+    # Explicit definition of the objective function for APGD
+    data_fid = 0.5 * pycdevu.SquaredL2Norm().asloss(data=measurements) * op
```
`as_loss`

joanrue commented on src/pycsou/_dev/fw_utils.py:
```diff
@@ -0,0 +1,122 @@
+        return y
+
+    @pycrt.enforce_precision(i="data")
+    def asloss(self, data: typ.Optional[pyct.NDArray] = None) -> pyco.ProxFunc:
```
`as_loss`

joanrue commented on src/pycsou/_dev/fw_utils.py:
```diff
@@ -0,0 +1,122 @@
+        y = arr / (2 * tau + 1)
+        return y
+
+    @pycrt.enforce_precision(i="data")
```
I think you need to add `allow_None=True` in order to be able to handle a `None` input.
`@pycrt.enforce_precision(i="data", allow_None=True)`

joanrue commented on src/pycsou/_dev/fw_utils.py:
```diff
@@ -0,0 +1,122 @@
+        return xp.clip(xp.abs(arr) - tau, a_min=0, a_max=None) * xp.sign(arr)
+
+    @pycrt.enforce_precision(i="data")
+    def asloss(self, data: typ.Optional[pyct.NDArray] = None) -> pyco.ProxFunc:
```
`allow_None=True`

joanrue commented on src/pycsou/opt/solver/fw_lasso.py:
```diff
@@ -0,0 +1,463 @@
+        self.data = pycrt.coerce(data)
+
+        self._data_fidelity = 0.5 * pycdevu.SquaredL2Norm().asloss(data=self.data) * self.forwardOp
+        self._penalty = self.lambda_ * pycdevu.L1Norm()
```
`as_loss`

joanrue reviewed (commented):
Impressive work!

Commit:
[abc.solver] Formatting/Cleanup.

Commit:
[doc] Change module paths due to use of from-style imports.

Commit:
[opt.solver] Update annotations.

Commit:
[compound] Update annotations.

Commit:
[operator.linop] Formatting/Cleanup.

Commit:
[operator][BugFix] Use function to gather core operators.

Commit:
[abc.operator] Cleanup/Formatting.

* Refactor Property class into Enumeration.
* specialize() is now asop().
* squeeze() is now _squeeze(). [Users don't need this.]
* arithmetic moved to separate module.

Commit:
[operator] Implement .asop()

Commit:
[operator][BugFix] _squeeze()-ing LinOps requires adding Proximal/Gradient properties.

Commit:
[operator][BugFix] LinFunc.from_array() must reshape 1D inputs to call super().

Commit:
[test_example] Fix Annotations

Commit:
[test_conftest] Drop argshift/argscale tests.

These are considered as operator arithmetic. As such their verification will be done as part of the arithmetic suite.

Commit:
[operator][BugFix] Override LinFunc implementations where needed.

Commit:
[test_conftest] Update all OpT test classes.

Commit:
[arithmetic] Implement scale()

Commit:
[arithmetic] Implement pow()

Commit:
[arithmetic] Implement argscale()

Commit:
[operator][BugFix] Squeezed LinOps must lose LINEAR_* properties.

Commit:
[operator][BugFix] Domain-Agnostic LinFuncs forbidden.

Reason: adjoint size undefined.

Commit:
[operator][BugFix] Handle `damp` correctly.

kwargs.pop("damp") fails given unspecified default value.

Commit:
[operator][BugFix] Drop double-defined properties().

Commit:
[solver][pgd][BugFix] Auto-infer NullFunc() dimension if f/g is domain-agnostic.

NullFunc can’t infer shape from other op.
Cheat: since adjoint not required in PGD, initialize NullFunc with dim=1

Commit:
[operator][BugFix] Avoid double-parentheses.

Commit:
[operator][BugFix] Allow squeezing LinOp to SquareOp when possible.

Commit:
[operator][BugFix] [co]gram() is a no-op for OrthProjOp.

Commit:
[operator] Enhance IdentityOp.

All methods forward to right superclass PosDefOp, OrthProjOp, or UnitOp.

Commit:
[operator.linop.base] Drop unused imports.

Commit:
[operator][BugFix] Redefine NullFunc in terms of NullOp.

Commit:
[operator] Specialize HomothethyOp.

Commit:
[operator.linop.base] Add DiagonalOp

Commit:
[test][operator.linop.base] IdentityOp()

Commit:
[test][operator.linop.base] NullOp()

Commit:
[test][operator.linop.base] LinFunc()

Commit:
[util] Fine-tune internal Pycsou warnings.

Situation before this commit: all warnings raised internally were UserWarnings. Consequence: cannot easily capture/filter them in the test suite.

Situation after this commit: introduce fine-tuned warnings for precision and dense-array related issues.

Commit:
[test][operator][BugFix] Do not allow casting to domain-agnostic linfuncs.

Commit:
[test][operator] Drop unnecessary Precision-Warnings.

Most have disappeared with the fine-tuning of the test suite.

Commit:
[test][BugFix] DiffMap: update use of ExplicitLinOp.

Commit:
[operator][BugFix] Let ExplicitLinOp() cast operator to the right Operator class itself.

Advantages:
* avoids having circular .asop() errors.
* conceptually more correct: if user asks for a certain class, we are not supposed to squeeze it for them.

Commit:
[test][operator][BugFix] Include complex-valued dtypes in isclose().

Required for test_value1D_eigvals() applied to Normal operators.

Commit:
[operator][BugFix] Set seed in svdvals() for deterministic tests.

Commit:
[hutchpp][BugFix] Computing via Dask returns incorrect values.

Reason: it seems Dask modifying `e` in-place causes overwrites which are not tracked well by dask.

Solution: recreate the canonical vector at each iteration.

Commit:
[arithmetic][pow] Negative powers are not supported.

Commit:
[operator.func] Formatting/Cleanup.

QuadraticFunc arithmetic removed. (Will be moved to dedicated arithmetic module.)

Commit:
[WIP] Arithmetic Notes

Commit:
[test][operator.linop.base] HomothetyOp()

Commit:
[test][operator.linop.base] DiagonalOp()

Commit:
[operator][from|to_sciop][BugFix] Force backend compliance.

Problem: SciOp's in/out signature is:
* in: np/cp
* out: np/cp

Dask inputs/outputs are not supported. As such any stateful operator returning dask arrays must be forced to evaluate prior to returning.

Commit:
[test][operator.linop.base] ExplicitLinOp()

Commit:
[arithmetic] Add base Rule hierarchy.

Commit:
[operator][map] Add ConstantValue().

Needed for arithmetic.ArgScaleRule().

Commit:
[operator][arithmetic] Use Rule classes instead of functions.

Commit:
[operator][ProxFunc] Refactor use of MethodType/partial.

Commit:
[operator][BugFix] has() is a classmethod.

Commit:
[linalg][BugFix] Missing import

Commit:
[operator][pinv] Use HomothethyOp directly for better performance.

(Also avoids operator inference issues due to IdentityOp's special status. [Does not fit in a single core category.])

Commit:
[operator][linfunc][BugFix] Transpose customization required.

Commit:
[test][operator.func] QuadraticFunc()

Commit:
[nullOp][BugFix] Gram/Cogram must be wrapped as PosDefOp.

Commit:
[solver][BugFix] Close log file-handlers at exit.

This bug only manifests itself when running 1000+ solver instances per Python interpreter. By not explicitly closing file-handlers, the OS quickly runs out of free file-descriptors to handle future solver instances.

Solution: close file-handlers when Solver has stopped.
For ASYNC/BLOCK-executed solvers: done in Solver.stop();
For MANUAL-executed solvers: done in Solver.steps().

Commit:
[operator][dagger] Simplify use of MethodType.

Commit:
[arithmetic] Implement ArgScaleRule.

Commit:
[arithmetic] Implement ChainRule.

Commit:
[arithmetic] Set default arithmetic attributes in Rule().

Commit:
[arithmetic] Implement PowerRule.

Commit:
[test][arithmetic] Initialize test package.

Commit:
[operator][BugFix][orthproj.pinv] Return solution in Range(Op) for consistency.

Commit:
[arithmetic] Enable Quadratic inference.

Would not work before since QuadraticFunc was not a core operator.
We fix the problem by making a _QuadraticFunc abstract class which gets included in core_operators().

Commit:
[arithmetic] Implement ScaleRule.

Commit:
[arithmetic][ScaleRule][BugFix] PROX-property must be preserved for LinFuncs.

Commit:
[arithmetic][ScaleRule][BugFix] .jacobian() is identity for LinOps.

Commit:
[test][arithmetic] ScaleRule.

Commit:
[test][arithmetic] ArgScale.

Commit:
[test][ArgScaleRule][BugFix] diff_lipschitz() grows quadratically.

Commit:
[arithmetic][ArgScaleRule][BugFix] .jacobian() is identity for LinOps.

Commit:
[test][arithmetic] ArgShift.

Commit:
[arithmetic] Implement ArgShiftRule.

Commit:
[arithmetic] Implement AddRule.

Commit:
arithmetic addrule wip

Commit:
[test][operator] Replace Rotation by Permutation to test D>3 case.

Commit:
[linop][identity] Drop inheritance from PosDef & OrthProj.

Reason: arithmetic rules were not designed to cope with operators present in multiple branches of the hierarchy.

Effect on IdentityOp properties: none.
Reason: being a special Operator, all meaningful methods are overwritten with efficient alternatives.

Commit:
[util][array_module] Add copy_if_unsafe().

Commit:
[util][array_module] Add read_only().

Commit:
[operator][BugFix] Force non-transparent functions to be read-only.

Commit:
[test][operator] Verify referential transparency.

Synopsis: some operator arithmetic methods (apply/adjoint/grad/prox) may return arrays which reference their inputs for efficiency.
Care must be taken in these scenarios before mutating the function outputs.

This commit adds a test to alert users when their implementations may suffer from side-effects.

Commit:
[operator|arithmetic][BugFix] Allow safe in-place updates.

Commit:
[test][operator] Replace CabinetProjection with Oblique to test D>3 case.

Commit:
[test][operator][BugFix] dim -> N rename for consistency with other test operators.

Commit:
[test][operator][BugFix][pinv|dagger] Protect against read-only asarray() outputs for special operators. (ex: NullOp)

Commit:
[test] Use Oblique/Permutation in place of (old) CabinetProjection/Rotation.

Commit:
[test][arithmetic] Run test_value1D_svdvals() at reduced precision.

Synopsis: when seeking the smallest-magnitude singular values via svdvals(), the iterative algorithm used converges to values close to the ground-truth (GT).
However, for 0-valued singular vectors, the relative error between converged solution and GT is sometimes slightly higher than the relative tolerance set for FP32/FP64 to consider them identical.

We therefore assess svdvals() outputs at FP32-precision only. This precision is enough to query an operators spectrum, or evaluate the conditioning number of a LinOp.

Commit:
[operator][BugFix][transpose] Force __call__ to be forwarded.

Special __*__ methods don't always forward automatically through .asop(). [Not Pycsou-specific: I have experienced this in pypeline as well.]

The lack of forwarding is clearly visible when doing operator arithmetic, so we add the (potentially-)redundant line to be safe.

Commit:
[arithmetic][AddRule][BugFix] LINEAR_NORMAL property is lost during additive composition.

Exception: the above property is preserved only if both operands are self-adjoint.

Commit:
[arithmetic][AddRule][BugFix] .jacobian() is identity for LinOps.

Commit:
[arithmetic][AddRule][BugFix] Range-broadcasted addition implies Lipschitz/Diff-Lipschitz/Adjoint pre/post-processing.

* (Diff-)Lipschitz constant of broadcasted operand must be magnified.
* Adjoint input of range-broadcasted operand must be summed.
* Lipschitz constant of (linear + linear) must by recomputed. (Reason: L1+L2 is an upper-bound; linops require lipschitz(svd) to return svd_max.)

Commit:
[arithmetic][AddRule][BugFix] Fail quickly if operand shapes incompatible.

Commit:
[test][arithmetic] ScaleRule: also test "/".

Commit:
[operator][BugFix] Must add DIFFERENTIABLE_FUNCTION property when squeezing DiffMaps.

Commit:
[arithmetic][ChainRule][BugFix] UNITARY implies NORMAL.

Commit:
[arithmetic][ChainRule][BugFix] QUADRATIC implies PROXIMABLE.

Commit:
[arithmetic][ChainRule][BugFix] LINEAR implies PROXIMABLE.

Commit:
[arithmetic][ChainRule] Before 1st test run.

Commit:
[operator][linop][BugFix] Favor DenseWarning over UserWarning for operations which revert to matrix-based ops.

Commit:
[DiagonalOp][BugFix] Fallback on universal function float() instead of backend-non-agnostic method `.item()`.

Dask arrays don't have .item(). Since Homothethy's are module/precision-agnostic, it is safe to cast constant to a CPU array.

Commit:
[test][arithmetic][setup] Silence DenseWarnings.

It is expected for DenseWarning to be raised when creating QuadraticFunc given small instance sizes.

Commit:
[arithmetic][ChainRule][BugFix] Lipschitz constant of (linear * linear) must by recomputed.

(Reason: L1*L2 is an upper-bound; linops require lipschitz(svd) to return svd_max.) [test_math3_lipschitz]

Commit:
[test][arithmetic][ChainRule][BugFix] Work around QuadraticFunc domain-size limitations.

Commit:
[arithmetic][ChainRule][BugFix] Remove reshape()-ambiguities due to domain-agnosticity.

Commit:
[arithmetic][ChainRule][BugFix] Diff-Lipschitz constant cannot be inferred for compositions where one term is non-linear.

Commit:
[test][arithmetic] Add.

Commit:
[test][operator] Provide some slack when comparing (L_svds <= L_fro).

For LinOps with svdmax <= 1, both terms can reach similar precision such that L_fro <= L_svds.

To avoid the issue, we use our own less_equal() function.

Commit:
[operator][linop][BugFix] `fro`-estimate must be quadratically shrunk.

We have \norm{A}_{F}^{2} = \tr(A^{T} A), hence hutchpp(op) returns the SQUARED Lipschitz constant.

This issue only becomes apparent for operators s.t. svdmax(A) < 1.

Commit:
[arithmetic][ChainRule][BugFix] Selectively preserve PROXIMABLE property for case (LinFunc \comp PROXIMABLE).

(LinFunc \comp PROXIMABLE) is proximable if (LinFunc(1,1) > 0).

Commit:
[operator][Quadratic] Use reduced-precision to compare [fenchel_]prox() outputs.

For the same reason we rely on reduced-precision comparison for [pinv|dagger]().

Commit:
[test][operator][BugFix] np.half is not a dtype: uniformize isclose() precision types.

All keys in the isclose() dictionary are dtypes, except np.half.
Replaced with np.dtype(np.half) for better clarity.

Commit:
[arithmetic][ChainRule][BugFix] Propagate proximal property for Quadratics.

Commit:
[arithmetic][Quadratic] Add _hessian() arithmetic rules.

(See comments in _QuadraticFunc() as to why this exists.)

Commit:
[operator][deprecation] Remove asloss().

asloss() is not an arithmetic method (and cannot become one due to operator-specific rule-set).
As such it cannot propagate through operator arithmetic.
This makes asloss() practically unuseable for even toy problems.

This commit entirely removes asloss() from pycsou's API.
No deprecation notice will be given.

Commit:
[operator][BugFix] Avoid accuracy loss for LinFunc.fenchel_prox().

Prior to this commit we rely on Moreau's identity to compute LinFunc.fenchel_prox().
But LinFunc.fenchel_prox() = LinFunc.grad(). [Closed form solution]

We now use the new closed-form to avoid accuracy loss due to un-needed arithmetic induced by Moreau's identity. [Fixes a precision bug uncovered in arithmetic.]

Commit:
[operator][moreau_envelope][BugFix] Compensate for dask.linalg.norm() return-precision bug.

Commit:
[operator][moreau_envelope] Use in-place computations where possible.

Commit:
[arithmetic][AddRule][BugFix] prox() can be evaluated for (Quadratic + Quadratic).

Commit:
[doc][BugFix] Correct some rendering issues.

Commit:
[SWE] Update working notes.

Commit:
[SWE] Provide better developer installation instructions.

It is now simpler to install Pycsou in developer mode.
Moreover, we expect users to use the meta-packages `complete_[no_]gpu` to install a well-rounded pycsou environment.

Drawback of current installation method (using conda): we install GPU packages even though a machine may not have a GPU.
`pycsou.util.deps.CUPY_ENABLED` will be set to False if a GPU is unavailable, even if GPU packages are installed.
Moreover GPU tests will hence not run.

Commit:
[arithmetic][ScaleRule][BugFix] Do not need to overload data_pinv[T]().

Can be auto-inferred from data_[apply|adjoint]().

Commit:
[arithmetic][ArgScaleRule][BugFix] Do not need to overload data_pinv[T]().

Can be auto-inferred from data_[apply|adjoint]().

Commit:
[arithmetic][AddRule][BugFix] Reduce obliquity factor to improve linop conditioning numbers.

[pinv|dagger](damp=None) tests fail due to ill-conditioned composite linops. To avoid the issue, we engineer the test matrices to be roughly well-behaved.

Commit:
[arithmetic][ChainRule][BugFix] Improve linop conditioning numbers.

[pinv|dagger](damp=None) tests fail due to ill-conditioned composite linops. To avoid the issue, we only test [pinv|dagger]() with dampening.

Commit:
[operator][moreau_envelope][BugFix] Protect againts read-only arrays.

Commit:
[ConstantValue] Save memory on [apply|grad]() calls.

Commit:
[runtime] Drop QUAD-precision.

Reason: unsupported on GPUs...

Commit:
[test][operator][BugFix] Do not compare floats via `==`.

Commit:
[CG][BugFix] Use simplified dual-step "residual" initialization.

Commit:
[util][array_module][BugFix] CuPy does not support writeable-flag.

CuPy arrays are considered read-only if they do not own their memory.

Commit:
[tox] Output better logs when running tests.

Commit:
[operator][BugFix] Add missing dependencies.

Commit:
[operator][svdvals][BugFix] `random_state` unsupported by CuPy.

Commit:
[test][operator][BugFix] Refactor _data_fenchel_prox() to be CuPy-compatible.

Commit:
[test][operator][BugFix] Make test_transparent() CuPy-compatible.

Commit:
[test][operator][normalop] Make base example fully XP-agnostic.

Commit:
[test][operator] Warn if functions are non-transparent to CuPy inputs.

Commit:
[operator][pinv|dagger][API-change] `damp` defaults to 0.

Case damp=None too problematic during tests.

Commit:
[runtime] Drop QUAD-precision.

Reason: unsupported on GPUs...

Commit:
[diagonalop][BugFix] [svd|eig]vals, trace() now work if `vec` is a Dask array.

Commit:
[test][sciop][BugFix] Protect pytest setup stage from ill-defined operators.

Ex: imagine an operator with some dask variable internally s.t. op.apply(np.ndarray) does not return an np.ndarray.
Then the entire _sciop test infrastructure breaks down.

We reduce the breakage space by going through .asarray() when possible.

Commit:
[operator][sciop][BugFix] Do not coerce LinearOperator return types too strictly.

This bug was discovered following commit[53e5a6f3ff37e5fe1d4d73ed97281186da6369b5].

Commit:
[test][diagonalop][BugFix] Disable math tests designed solely for the numpy backend.

Previously DiagonalOp was only tested with a NumPy parametrization.
Extending the internal state vector to dask arrays uncovers these whitelisted bugs.

Commit:
[test][diagonalop][BugFix] Disable [svd|eig]vals() impossible parametrizations.

Commit:
[test][diagonalop][BugFix] Disable [from|to]_sciop() tests: too difficult to assess with internal state.

Commit:
[diagonalop][BugFix] asarray() now works with xp=CuPy.

Commit:
[diagonalop] Improve pinv/dagger performance.

Commit:
[explicitlinop] Rename encapsulated matrix. (Similar to DiagonalOp.)

Commit:
[test][operator][eigvals] Refactor width override.

No longer necessary to enumerate the widths since we only have SINGLE/DOUBLE.

Commit:
[test][explicitlinop][BugFix] without GPU array_initializers, everything works.

Commit:
[explicitlinop][lipschitz][BugFix] Now CuPy-compatible.

Commit:
[explicitlinop][asarray][BugFix] Now CuPy-compatible.

Commit:
[explicitlinop][trace][BugFix] Now CuPy-compatible.

Commit:
[test][explicitlinop][BugFix] data_apply must use domain-specific `_matrix`

Commit:
[test][explicitlinop] Reduce accepted sparse arrays to COO-like.

Commit:
[explicitlinop][API] Rename underlying array to public.

Commit:
[explicitlinop][API] Make class private.

Users are not expected to initialize _ExplicitLinOp via .__init__().
Instead privilege LinOp.from_array() as follows:
    A = <some user-specified positive-definite array>
    op = pyca.PosDefOp.from_array(A)
Reason: allows overloaded methods from specific subclasses to kick in for improved performance.

Commit:
[explicitlinop][asarray] Improve performance on CuPy arrays.

Avoid unnecessary copy to main memory.

Commit:
[operator][BugFix] Customize svdvals() for better performance, in particular for CuPy arrays.

Commit:
[operator][linop][BugFix] Correct import path for CuPy backend.

Commit:
[operator][linop][BugFix] Use svd() in place of svdvals().

CuPy has no function named svdvals(). svd/svdvals() differ only in their treatment of the degenerate empty-array (source: SciPy docs), hence we can safely call svd(_, compute_uv=False) as a substitute for dense evaluations.

Commit:
[operator][linop][BugFix] eig()/svd(): reduce dense evaluation threshold.

The prior threshold to resort to dense evaluation was k~=dim/codim.
We reduce it to half the shape. This allows us to sidestep corner cases where CuPy.eigh/svdvals are buggy.

Commit:
[test][normalop][eigvals] Detect expected failure if gpu=True.

# Problem: cupyx.scipy.sparse.linalg.eigs does not exist.
# Solution: expect an exception when importing it. Once CuPy add eigs(), then the test will fail and we will know that NormalOp.eigvals() now works on GPUs.

Commit:
[test][linop] Mask expected DenseWarning(s) on some small tests.

DenseWarning is emitted by many LinOp functions which fallback on dense computations in place of sparse ones depending on method parameters.

It is safe to mask them entirely during testing.

Commit:
[API][util.deps] Replace array_backend_info() with Enums() + Enum-methods.

Reasons:
* avoid string name errors in functions relying on redirect().
* no need to import libraries in modules which require deps.py: Enum methods can return array type or module as needed.

Commit:
[API][test] Extend isclose|allclose|less_equal to work with scalars.

Commit:
[test] Code formatting + moving definitions.

Commit:
[test][operator] Refactor to work with new fixture structure.

Added: exhaustive math tests done for all backend/precision pairs.

Commit:
[test][examples] Updated with new fixture structure.

Commit:
[test][operator][BugFix] Flag [eig|svdvals] limitation to `which=LM` (CuPy-only).

Commit:
[test][operator][BugFix] Flag [Normal|Unit]Op.eigvals() limitation. (CuPy-only)

Commit:
[operator][map] Refactored ConstantValued().

Commit:
[operator][linop] Refactor Null[Op|Func]().

Commit:
[operator][linop] Refactor IdentityOp().

Commit:
[operator][linop] Refactor HomothethyOp().

Commit:
[operator][linop] Refactor DiagonalOp().

Commit:
[test][operator][BugFix] Specify operator backend. (Needed for dask-eval.)

Commit:
[test][operator][BugFix] Safe instantiation for backend-specific operators.

Calling .gram() then .asarray() redirects calls to the generic PosDefOp.asarray() implementation, which may not work if `op` is backend-specific.

Solution: computing the Gram matrix manually solves the problem.

Commit:
[operator][linop] Refactor _ExplicitLinOp().

Commit:
[operator][linop][BugFix] Inform user of potential CuPy-specific bug.

Can't do anything in Pycsou about it: must be fixed upstream.

Commit:
[math][linalg] Use pycsou.util.deps tools in place of manual module checks.

Commit:
[operator][sciop][BugFix] Simplified due to operator test suite refactor.

Undoes commit[9dc4c85dd9424a529bc0d4a89a3ca7e484e5b44e]

Commit:
[operator] Use pycsou.util.deps tools in place of manual module checks.

Commit:
[arithmetic][AddRule] Refactor for readability.

Commit:
[test][arithmetic] Update to use `spec()` in place of manual `op/xp/width` specification.

Commit:
[operator][linop][BugFix] Forward .asarray() for transposed operators.

An operator may have defined a custom .asarray() method.
In such cases we want op.T to use it in place of the default implementation.

Commit:
[operator] Drop unused imports + code formatting + fix math/docs

Commit:
[arithmetic][LinOp][BugFix+Enhancement] Promote some helper methods to arithmetic methods.

Synopsis
--------
Before this commit, the only arithmetic property of the LinOp hierarchy was .adjoint(): all other helper methods rely on .apply()/.adjoint() to compute their values, ex:

* asarray()
* svdvals()
* eigvals() [NormalOp+]
* trace() [SquareOp+]
* pinv() / dagger()

If operators are backend-agnostic, then the default implementations of these listed methods work after any operator arithmetic -> nothing to do.

However, if the operator is domain-specific (or one of it's methods make it domain-specific), then the default implementations of listed methods may break during arithmetic composition.

Concrete Example of Broken Arithmetic
-------------------------------------

>>> op = 2 * <some domain-agnostic LinOp> = 2 * op_orig
>>> J = op.jacobian(<CuPy array>)  # domain-specific: J.apply/adjoint/... only accept CuPy inputs
>>> A = J.asarray(xp=np, dtype=<any>)  # ERROR

Reason: J.asarray() calls the default implementation of .asarray() based on J.apply(xp.eye(...)).
But J is domain-specific: the call to J.apply() will only work if `xp=cp`.
This violates the .asarray() signature where any value for `xp` should work.

If .asarray() was an arithmetic method however, then we could force the computation to be done via
    J.asarray(xp, dtype) = op_orig.asarray(xp, dtype) * 2,
a result which will work regardless of the value of `xp` since op_orig.asarray() can compute it without any problem.

Alternatively, for some arithmetic operators, a more-efficient algorithm may be used to evaluate a method.

Concrete Example of Faster (Non-Default) Implementation
-------------------------------------------------------

>>> op = -2 * <some domain-agnostic UnitOp> = 2 * op_orig
>>> D = op.svdvals(k=1, which='SM')  # works, but sub-optimal

op.svdvals() is suboptimal because `op` is just a NormalOp, hence piggy-backs on LinOp.svdvals(), an iterative algorithm.
However, we know from arithmetic rules that op.svdvals() could be computed more efficiently by piggy-backing on UnitOp.svdvals() which can be computed in closed form.
    op.svdvals(k) = op_orig.svdvals(k) * abs(-2)

Bottom line:
* some methods can be computed FASTER by inclusion in arithmetic rules, and
* some methods can only be computed CORRECTLY by inclusion in arithmetic rules.

Commit:
[test][operator][BugFix] allclose() needs a dtype parameter, not Width.

Commit:
[arithmetic][BugFix] Deepcopy of Python modules illegal.

Shallow copy sufficient since to mutable state is to be sent to **kwargs.

Commit:
[operator][LinFunc][asarray] Decrease memory requirements via in-place updates.

Commit:
[arithmetic][AddRule][BugFix] trace() should handle case SquareOp() + LinFunc()

LinFunc() does not have a .trace() method, so we rely on the (fast) .asarray() method pre-filled for this class.

Commit:
[arithmetic][ChainRule][BugFix] trace() requires explicit `self` argument.

Commit:
[test][operator] Verify double-transpose returns operator of same shape.

Commit:
[operator] Provide .transpose() in addition to Property[T].

Commit:
[operator][dagger][BugFix] Overwrite __call__ to avoid mis-interpretation during arithmetic.

Commit:
[operator][BugFix] Use `xp` to be backend-agnostic.

Commit:
[operator] Drop .dagger() from arithmetic.

Commit:
[arithmetic] Provide default implementation of arithmetic methods at top-level Rule class.

Commit:
[test][arithmetic] Add rectangular test cases.

Commit:
[operator][BugFix] .[co]gram() must return SelfAdjointOp or LinFunc.

Also drop cogram() override for UnitOp: uses NormalOp.cogram() by default, which points to self.gram().

Commit:
[operator][BugFix] .transpose() now works during operator arithmetic.

Commit:
[arithmetic] Propagate [co]gram().

Care must be taken in AddRule due to range-broadcasting.

Commit:
[quadratic][BugFix] Use Homothethy in place of a scaled-IdentityOp.

Commit:
[linop] Move IdentityOp to OrthProjOp.

Reason: required to avoid custom arithmetic rules just for the sole Operator belonging to multiple classes simultaneously.

Commit:
[operator] Add .expr().

We need a way to view what arithmetic operators do behind the scenes.
Operator.expr() allows one to print the arithmetic tree to debug problems.

Commit:
[doc][operator] Fix rendering issue.

Commit:
[test][operator][BugFix] Add .expr() to list of expected methods.

Commit:
[operator] Add .expr() support to ProxFunc.moreau_envelope().

Commit:
[operator][pinv|dagger] Simplify rules given ScaleRule properties.

Commit:
[operator] Add .expr() support to LinOp.dagger().

Commit:
[operator][dagger][BugFix] Return SquareOp when possible.

Commit:
[operator] Add .expr() support to LinOp.transpose().

Commit:
[operator] Add .expr() support to LinOp.[co]gram().

AdriaJ committed:
Merge pull request #7 from matthieumeo/v2-dev_kashani

V2 dev kashani

AdriaJ committed:
Adapt example with recent changes of the API. Working example.

AdriaJ committed:
Merge pull request #8 from matthieumeo/v2-dev

V2 dev

AdriaJ committed:
Merge branch 'v2-dev' into v2-fw

# Conflicts:
#	src/pycsou/abc/arithmetic.py
#	src/pycsou/abc/operator.py
#	src/pycsou/operator/func/__init__.py
#	src/pycsou/operator/linop/base.py
#	src/pycsou/operator/map/base.py
#	src/pycsou/opt/solver/pgd.py
#	src/pycsou/util/deps.py
#	src/pycsou/util/warning.py
#	src/pycsou_tests/operator/arithmetic/test_addrule.py
#	src/pycsou_tests/operator/arithmetic/test_argscalerule.py
#	src/pycsou_tests/operator/arithmetic/test_argshiftrule.py
#	src/pycsou_tests/operator/arithmetic/test_chainrule.py
#	src/pycsou_tests/operator/arithmetic/test_scalerule.py
#	src/pycsou_tests/operator/conftest.py
#	src/pycsou_tests/operator/examples/test_difffunc.py
#	src/pycsou_tests/operator/examples/test_func.py
#	src/pycsou_tests/operator/examples/test_proxdifffunc.py
#	src/pycsou_tests/operator/examples/test_proxfunc.py
#	src/pycsou_tests/operator/linop/base/test_explicitlinop.py
#	src/pycsou_tests/operator/map/base/test_constantvalued.py

AdriaJ committed:
Example of positivity constraint with the LASSO problem, PDS not working properly.

AdriaJ committed:
Quick changes of the solver, after recent changes on the arithmetic of the operators.

AdriaJ committed:
Quick changes of the solver, after recent changes on the arithmetic of the operators.

AdriaJ committed:
Integrate a flag for a positivity constraint in the FW solvers.

AdriaJ committed:
Specify shape of the operators in the objective function of the LASSO.

AdriaJ committed:
Merge branch 'v2-dev' of https://github.com/matthieumeo/pycsou into matthieumeo-v2-dev

# Conflicts:
#	src/pycsou/opt/solver/pds.py

AdriaJ committed:
Merge branch 'v2-dev' of https://github.com/matthieumeo/pycsou into matthieumeo-v2-dev

# Conflicts:
#	src/pycsou/opt/solver/pds.py

AdriaJ committed:
Merge branch 'v2-dev' of https://github.com/matthieumeo/pycsou into matthieumeo-v2-dev

# Conflicts:
#	src/pycsou/opt/solver/pds.py

AdriaJ committed:
Merge branch 'v2-dev' of https://github.com/matthieumeo/pycsou into matthieumeo-v2-dev

# Conflicts:
#	src/pycsou/opt/solver/pds.py

AdriaJ committed:
Merge branch 'matthieumeo-v2-dev' into v2-dev

AdriaJ committed:
Merge branch 'v2-dev' into v2-fw

# Conflicts:
#	src/pycsou/opt/solver/pds.py

AdriaJ committed:
Merge branch 'v2-dev' into v2-fw

# Conflicts:
#	src/pycsou/opt/solver/pds.py

AdriaJ committed:
Merge branch 'v2-dev' into v2-fw

# Conflicts:
#	src/pycsou/opt/solver/pds.py

AdriaJ committed:
Debug the positivity constraint for the LASSO with APGD: proximal operator for l` plus indicator function of the non-negative orthant.

AdriaJ committed:
Positivity constraint added to the partial correction.

AdriaJ committed:
Merge pull request #10 from matthieumeo/v2-dev

V2 dev

AdriaJ committed:
Merge branch 'v2-dev' into v2-fw

# Conflicts:
#	src/pycsou/operator/func/__init__.py

AdriaJ committed:
Merge branch 'v2-dev' of https://github.com/matthieumeo/pycsou into v2-dev

AdriaJ committed:
Merge branch 'v2-dev' of https://github.com/matthieumeo/pycsou into v2-dev

AdriaJ committed:
Merge pull request #11 from matthieumeo/v2-dev

V2 dev

AdriaJ committed:
Merge branch 'v2-dev' into v2-fw

AdriaJ committed:
Update FW implementation with current state of Pycsou, solve correction step.

AdriaJ committed:
Add post_processing method.

AdriaJ committed:
Try some post processing in the example.

AdriaJ committed:
Merge branch 'v2-dev' of https://github.com/matthieumeo/pycsou into v2-dev

AdriaJ committed:
Merge branch 'v2-dev' into v2-fw

AdriaJ committed:
Reformat rs_correction and add min number of correction steps.

AdriaJ committed:
Add the possibility to bypass the computation of the diff_lipschitz_constant at .fit().
